// inversify-logger-middleware/es/serializers/text/text_serializer_utils.js
import { yellow, green, red } from "chalk";
var tree = {
    item: "└──"
};
var indentationStr = "    ";
function getIndentationForDepth(depth) {
    var indentationForDepth = indentationStr;
    for (var i = depth; i--; i > 0) {
        indentationForDepth = "" + indentationForDepth + indentationStr;
    }
    return indentationForDepth;
}
function makePropertyLogger(indentationForDepth) {
    return function (textEntry, tabCount, key, value) {
        var line = "" + textEntry + indentationForDepth;
        for (var i = tabCount; i > 0; i--) {
            line = "" + line + indentationStr;
        }
        line = "" + line + tree.item + " " + key;
        if (value !== undefined) {
            if (typeof value !== "boolean") {
                var val = value ? value.toString().split("\n").join("") : "null";
                line = line + " : " + yellow(val);
            }
            else {
                line = line + " : " + yellow(value.toString());
            }
        }
        return line + "\n";
    };
}
export { getIndentationForDepth, makePropertyLogger, yellow, green, red };
